docs/jax-101/07-state.md

```diff
@@ -0,0 +1,283 @@
+
+*Authors: Vladimir Mikulik*
+
+This section explores how JAX constrains the implementation of stateful programs.
+
++++ {"id": "Avjnyrjojo8z"}
+
+## Motivation
+
+In machine learning, program state most often comes in the form of
+* model parameters,
+* optimizer state, and
+* stateful layers, such as [BatchNorm](https://en.wikipedia.org/wiki/Batch_normalization).
+
+Some JAX transformations, most notably `jax.jit`, impose constraints on the functions they transform. In particular, the function transformed by `jax.jit` must have no side-effects. This is because any such side-effects will only be executed once, when the python version of the function is run during compilation. These side-effects will not be executed by the compiled function on subsequent runs.
+
+Changing program state is one kind of side-effect. So, if we can't have side effects, how do we update model parameters, the optimizer state, and use stateful layers in our models? This colab will explain this in detail, but the short answer is: with [functional programming](https://en.wikipedia.org/wiki/Functional_programming).
+
+
+```{code-cell} ipython3
+:id: S-n8XRC2kTnI
+
+import jax
+import jax.numpy as jnp
+```
+
++++ {"id": "s_-6semKkSzp"}
+
+
+
+## A simple example: Counter
+
+Let's start by looking at a simple stateful program: a counter.
+
+```{code-cell} ipython3
+---
+id: B3aoCHpjg8gm
+outputId: 5cbcfbf5-5c42-498f-a175-050438518337
+---
+class Counter:
+  """A simple counter."""
+
+  def __init__(self):
+    self.n = 0
+
+  def count(self) -> int:
+    """Increments the counter and returns the new value."""
+    self.n += 1
+    return self.n
+
+  def reset(self):
+    """Resets the counter to zero."""
+    self.n = 0
+
+
+counter = Counter()
+
+for _ in range(3):
+  print(counter.count())
+```
+
++++ {"id": "SQ-RNLfdiw04"}
+
+The `n` attribute maintains the counter's _state_ between successive calls of `count`. It is modified as a side effect of calling `count`.
+
+Let's say we want to count fast, so we `jax.jit` the `count` method. (In this example, this wouldn't actually help speed anyway, for many reasons, but treat this as a toy model of wanting to jit-compile the update of model parameters, where `jax.jit` makes an enormous difference).
+
+```{code-cell} ipython3
+---
+id: 5jSjmJMon03W
+outputId: d952f16b-9b30-4753-ed94-cc914a929a36
+---
+counter.reset()
+fast_count = jax.jit(counter.count)
+
+for _ in range(3):
+  print(fast_count())
+```
+
++++ {"id": "weiI0V7_pKGv"}
+
+Oh no! Our counter isn't working. This is because the line
+```
+self.n += 1
+```
+in `count` was only called once, when JAX was compiling the method call. Moreover, since the return value doesn't depend on the arguments to `count`, once it returned the first 1, subsequent calls to `fast_count` will always return 1. This won't do. So, how do we fix it?
```

Comment:
Suggestion: nit (present tense, maybe better formatting)

```suggestion
in `count` was only called once, when JAX was compiling the method call. Moreover, since the return value doesn't depend on the arguments to `count`, once it returns the first `1`, subsequent calls to `fast_count` will always return `1`. This won't do. So, how do we fix it?
```